truss/templates/server/common/schema.py

```diff
@@ -0,0 +1,164 @@
+    parameter_types = list(input_parameters.values())
+
+    if len(parameter_types) > 1:
+        return None
+
+    input_type = parameter_types[0].annotation
+
+    if (
+        input_type == inspect.Signature.empty
+        or not isinstance(input_type, type)
+        or not issubclass(input_type, BaseModel)
+    ):
+        return None
+
+    return input_type
+
+
+def _parse_output_type(output_annotation: Any) -> Optional[Tuple[Optional[type], bool]]:
+    """
+    Therea are 4 possible cases for output_annotation:
+    1. Data object -- represented by a Pydantic BaseModel
+    2. Streaming -- represented by a Generator or AsyncGenerator
+    3. Async -- represented by an Awaitable
+    4. Streaming or data object -- it is possible for a function to return both
+       a data object and a streaming object. This is represented by a Union of
+       a Pydantic BaseModel and a Generator or AsyncGenerator
+
+    If the output_annotation does not match one of these cases, returns None
+    """
+    if isinstance(output_annotation, type) and issubclass(output_annotation, BaseModel):
+        return output_annotation, False
+
+    if _is_generator_type(output_annotation):
+        return None, True
+
+    if _is_awaitable_type(output_annotation):
+        output_type = retrieve_base_class_from_awaitable(output_annotation)
+        if not output_type:
+            return None
+        return output_type, False
+
+    if _is_union_type(output_annotation):
+        output_type = retrieve_base_class_from_union(output_annotation)
+        if not output_type:
+            return None
+        return output_type, True
+
+    return None
+
+
+def _is_generator_type(annotation: Any) -> bool:
+    base_type = get_origin(annotation)
+    return isinstance(base_type, type) and issubclass(
+        base_type, (Generator, AsyncGenerator)
+    )
+
+
+def _is_union_type(annotation: Any) -> bool:
+    return get_origin(annotation) == Union
+
+
+def _is_awaitable_type(annotation: Any) -> bool:
+    base_type = get_origin(annotation)
+    return isinstance(base_type, type) and issubclass(base_type, Awaitable)
+
+
+def retrieve_base_class_from_awaitable(awaitable_annotation: type) -> Optional[type]:
+    """
+    Returns the base class of an Awaitable type if it is of the form:
+    Awaitable[PydanticBaseModel]
+
+    Else, returns None
+    """
+    (awaitable_type,) = get_args(
+        awaitable_annotation
+    )  # Note that Awaitable has only one type argument
+    if isinstance(awaitable_type, type) and issubclass(awaitable_type, BaseModel):
+        return awaitable_type
+
+    return None
+
+
+def _extract_pydantic_base_models(union_args: tuple) -> list:
+    """
+    Extracts any pydantic base model arguments from the arms of a Union type.
+    The two cases are:
+    1. Union[PydanticBaseModel, Generator]
+    2. Union[Awaitable[PydanticBaseModel], AsyncGenerator]
+    So for Awaitables, we need to extract the base class from the Awaitable type
```

Comment:
Could we add to the docstring what this would return in these two cases? That might help clarify that each element of the returned list corresponds to a pydantic model / awaitable arg, and none of the elements correspond to generator args

Reply:
Rather than add to the docstring, I just made the return type of this function more specific.